### DISCUSSION TITLE
Build tasmota for 2MB flash

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#12298] [Q&A] [answered]
I want to build a tasmota image for a shelly dimmer 2 which has apparently 2MB flash.

At this moment I build a 1MB flash image and for upgrades I first have to flash the minimal version before flashing the update because the image is around 600kb.

I found options to build for 1NB or 4MB flash but not for 2 MB flash.

How cab I build a 2MB (0MB spiffs) version?

**Comment by sfromis** [comment, marked as answer]
See [platformio_override_sample.ini](https://github.com/arendst/Tasmota/blob/development/platformio_override_sample.ini) where you can set `board = esp8266_2M1M` - you do not get more than 1 MB for firmware, as this is what ESP8266 supports.

Always use the .gz binary for OTA upgrading, reduces the need for using -minimal. You could also setup a local http server for your upgrading, allowing Tasmota to *automatically* use minimal when relevant.

Also see definitions in https://github.com/arendst/Tasmota/tree/development/boards

**Reply by Jason2866** [reply]
A 2M no SPIFFS build makes no sense with Tasmota. The borders we are hitting is NOT running out of flash space for code and OTA with 1M flash space. The limiting factor when using a lot of drivers is running out of free RAM. Anyway for a Shelly dimmer 1M is more than enough.

**Reply by sfromis** [reply]
When the desire is to avoid minimal for OTA, I was wondering how it would work out modifying `esp8266_4M2M` which is "1MB firmware, 1MB OTA, 2MB filesystem" to chop off the file system, and still have the OTA partition on a 2M flash board?

**Reply by Jason2866** [reply]
> When the desire is to avoid minimal for OTA, I was wondering how it would work out modifying `esp8266_4M2M` which is "1MB firmware, 1MB OTA, 2MB filesystem" to chop off the file system, and still have the OTA partition on a 2M flash board?

Use linker script `eagle.flash.2m.ld` in the boards file. This will generate a version with 1M code space and 1M OTA space

**Reply by jeroenst** [reply]
Ok thank you, I will give that a try!

**Comment by dkadioglu** [comment]
I hope, it is okay to chime in.
I am experimenting with building a firmware for 2M256, which finishes successfully and runs. However, what I don't understand is the difference in size compared to standard [env:tasmota]:
tasmota.bin 683.2k
tasmota2M256.bin 731.3k


What I did is simply adding the following to platformio_tasmota_cenv.ini:
[env:tasmota2M256]
board                   = esp8266_2M256

I would have expected, that there is no size difference. If the size difference is to be expected, could someone explain why this happens?

**Reply by sfromis** [reply]
AFAIK, esp8266_2M256 includes a file system and code to support this

**Reply by dkadioglu** [reply]
Thanks for this explanation. However, how can I achieve the same size, e.g. reserving the second megabyte for OTA?
I tried to build a firmware based on a newly created board file:
```
{
  "build": {
    "arduino": {
      "ldscript": "eagle.flash.2m.ld"
    },
    "core": "esp8266",
    "extra_flags": "-DESP8266 -DARDUINO_ARCH_ESP8266 -DARDUINO_ESP8266_ESP01 -DESP8266_2M",
    "f_cpu": "80000000L",
    "f_flash": "40000000L",
    "flash_mode": "dout",
    "mcu": "esp8266",
    "variant": "generic"
  },
  "connectivity": [
    "wifi"
  ],
  "frameworks": [
    "arduino"
  ],
  "name": "Espressif Generic ESP8266 Tasmota 1M sketch NO FS",
  "upload": {
    "maximum_ram_size": 81920,
    "maximum_size": 995326,
    "require_upload_port": true,
    "resetmethod": "ck",
    "speed": 115200
  },
  "url": "http://www.esp8266.com/wiki/doku.php?id=esp8266-module-family",
  "vendor": "Espressif"
}
```
Still, the resulting image is 731,3k. Am I missing something to disable the filesystem?

**Reply by Jason2866** [reply]
Why? There is no advantage. There is a high possibility that you run in issues when you flash a standard firmware. We do not recommend!

**Reply by dkadioglu** [reply]
I just had the idea, to make full use of the size of the flash chip. As I don't need the filesystem, I thought great, let's use the space for OTA to not be required to use minimal anymore for updates. If that is not recommended, accepted. However, just to clarify, is it ok to use 2M256?

**Reply by Jason2866** [reply]
Not recommended nor needed so not tested, idk what happens. You walk alone.
Again you dont need minimal with the standard 1M build. Just use the `*.gz` firmware to update.

**Reply by dkadioglu** [reply]
Ok, thanks for clarifying. Totally understandable, that not everything can be tested nor supported!

**Comment by Subnum12** [comment]
Actually I run now this board file on 2M devices:
```{
  "build": {
    "arduino": {
      "ldscript": "eagle.flash.2m1m.ld"
    },
    "core": "esp8266",
    "extra_flags": "-DARDUINO_TASMOTA -DESP8266 -DARDUINO_ARCH_ESP8266 -DARDUINO_ESP8266_ESP01 -DESP8266_1M",
    "f_cpu": "80000000L",
    "f_flash": "40000000L",
    "flash_mode": "dout",
    "mcu": "esp8266",
    "variant": "generic"
  },
  "connectivity": [
    "wifi"
  ],
  "frameworks": [
    "arduino"
  ],
  "name": "Espressif Generic ESP8266 Tasmota 1M sketch NO FS",
  "upload": {
    "maximum_ram_size": 81920,
    "maximum_size": 995326,
    "require_upload_port": true,
    "resetmethod": "ck",
    "speed": 115200
  },
  "url": "http://www.esp8266.com/wiki/doku.php?id=esp8266-module-family",
  "vendor": "Espressif"
}
```
or on 4M systems just `"ldscript": "eagle.flash.4m2m.ld"`

@Jason2866 mentioned: 
>Why? There is no advantage. You dont need minimal with the standard 1M build. Just use the *.gz firmware to update.

Actually I made different experience. 
1. with 1M you cannot use the *.gz firmware to update via OTA without minimal -> Not enough space
2. and more important for me : Not having file system is required for using define `USE_SCRIPT`.  If you want `USE_SCRIPT` you cannot have file system as scripts than requires not 1k but 10k heap. That means only about 13k left. Therefore tasmotta-4M with `USE_SCRIPT ` makes no sense.

With board setting above OTA will work and with define `USE_SCRIPT` you will still have about 22k left.

**Reply by sfromis** [reply]
It is certainly *possible* to have a file system and scripter at the same time, but you may have to trim the many available build time options available with the scripter, to make it workable when having a file system. You just can't have it all at the same time.

The "unofficial" scripter build setup includes a lot of optional features and default values for memory-related settings. You can selectively tweak to shoehorn things in.
https://github.com/Jason2866/Tasmota-build/blob/special_builder/tasmota/user_config_override.h#L1620-L1934

Of course, you may need no file system, but having that is not impossible with the scripter.
